pkg/lightning/backend/local/iterator.go

```diff
@@ -0,0 +1,193 @@
+		return false
+	}
+	d.fill()
+	return d.err == nil
+}
+
+func (d *duplicateIter) First() bool {
+	if d.err != nil || !d.iter.First() {
+		return false
+	}
+	d.fill()
+	return d.err == nil
+}
+
+func (d *duplicateIter) Last() bool {
+	if d.err != nil || !d.iter.Last() {
+		return false
+	}
+	d.fill()
+	return d.err == nil
+}
+
+func (d *duplicateIter) fill() {
+	d.curKey, _, _, d.err = d.keyAdapter.Decode(d.curKey[:0], d.iter.Key())
+	d.curRawKey = append(d.curRawKey[:0], d.iter.Key()...)
+	d.curVal = append(d.curVal[:0], d.iter.Value()...)
+}
+
+func (d *duplicateIter) flush() {
+	d.err = d.writeBatch.Commit(pebble.Sync)
+	d.writeBatch.Reset()
+	d.writeBatchSize = 0
+}
+
+func (d *duplicateIter) record(key []byte, val []byte) {
+	d.engineFile.Duplicates.Inc()
+	d.err = d.writeBatch.Set(key, val, nil)
+	if d.err != nil {
+		return
+	}
+	d.writeBatchSize += int64(len(key) + len(val))
+	if d.writeBatchSize >= maxDuplicateBatchSize {
+		d.flush()
+	}
+}
+
+func (d *duplicateIter) Next() bool {
+	recordFirst := false
+	for d.err == nil && d.ctx.Err() == nil && d.iter.Next() {
+		d.nextKey, _, _, d.err = d.keyAdapter.Decode(d.nextKey[:0], d.iter.Key())
+		if d.err != nil {
+			return false
+		}
+		if !bytes.Equal(d.nextKey, d.curKey) {
+			d.curKey, d.nextKey = d.nextKey, d.curKey[:0]
+			d.curRawKey = append(d.curRawKey[:0], d.iter.Key()...)
+			d.curVal = append(d.curVal[:0], d.iter.Value()...)
+			return true
+		}
+		log.L().Debug("duplicate key detected", logutil.Key("key", d.curKey))
+		if !recordFirst {
+			d.record(d.curRawKey, d.curVal)
+			recordFirst = true
+		}
+		d.record(d.iter.Key(), d.iter.Value())
+	}
+	if d.err == nil {
+		d.err = d.ctx.Err()
+	}
+	return false
+}
+
+func (d *duplicateIter) Key() []byte {
+	return d.curKey
+}
+
+func (d *duplicateIter) Value() []byte {
+	return d.curVal
+}
+
+func (d *duplicateIter) Valid() bool {
+	return d.err == nil && d.iter.Valid()
+}
+
+func (d *duplicateIter) Error() error {
+	return multierr.Combine(d.iter.Error(), d.err)
+}
+
+func (d *duplicateIter) Close() error {
+	if d.err == nil {
+		d.flush()
+	}
+	d.writeBatch.Close()
+	return d.iter.Close()
+}
+
+func (d *duplicateIter) OpType() sst.Pair_OP {
+	return sst.Pair_Put
+}
+
+var _ kv.Iter = &duplicateIter{}
+
+func newDuplicateIter(ctx context.Context, engineFile *File, opts *pebble.IterOptions) kv.Iter {
+	newOpts := &pebble.IterOptions{TableFilter: opts.TableFilter}
+	if len(opts.LowerBound) > 0 {
+		newOpts.LowerBound = codec.EncodeBytes(nil, opts.LowerBound)
+	}
+	if len(opts.UpperBound) > 0 {
+		newOpts.UpperBound = codec.EncodeBytes(nil, opts.UpperBound)
+	}
+	return &duplicateIter{
+		ctx:        ctx,
+		iter:       engineFile.db.NewIter(newOpts),
+		engineFile: engineFile,
+		keyAdapter: engineFile.keyAdapter,
+		writeBatch: engineFile.duplicateDB.NewBatch(),
+	}
+}
+
+func newNormalKeyIter(ctx context.Context, engineFile *File, opts *pebble.IterOptions) kv.Iter {
```

Comment:
this method name is some what misleading 😲， how about `newKVIter`

Reply:
okay